executor: Add Global to SHOW INDEX output | tidb-test=pr/2391

samba-rgb opened the pull request:
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. show.go, show_test.go, planbuilder.go
2. added Global to show indexes

-->

### What problem does this PR solve?
No info about index if its global or not , added in show index in this pr


Issue Number: ref #55452

Problem Summary:

### What changed and how does it work?

added Global index from index info

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

lance6716 commented:
/ok-to-test

mjonss commented on pkg/executor/show.go:
```diff
@@ -805,6 +807,14 @@ func (e *ShowExec) fetchShowIndex() error {
+		isPartitioned := pi != nil && pi.Type != pmodel.PartitionTypeNone
+
+		if isPartitioned && e.Ctx().GetSessionVars().EnableGlobalIndex {
+			isGlobalIndex = "YES"
```
If PKIsHandle, then it is by definition a LOCAL index, i.e. each partition will have the clustered index, and the PK column must be a part of the partitioning expression.

Test by:
```sql
set tidb_enable_global_index = 0;
create table t (a int primary key clustered, b int) partition by hash (a) partitions 2;
show create table t;
show index from t;
set tidb_enable_global_index = 1;
show index from t;
```

mjonss commented on pkg/executor/show.go:
```diff
@@ -822,6 +832,7 @@ func (e *ShowExec) fetchShowIndex() error {
 			"YES",            // Index_visible
 			nil,              // Expression
 			"YES",            // Clustered
+			isGlobalIndex,    // Global_index
```
So here you can just use "NO"
```suggestion
			"NO",    // Global_index
```

mjonss replied:
Ah, sorry, I had not indented it correctly :(

mjonss commented on pkg/executor/show_test.go:
```diff
@@ -97,3 +97,32 @@ func TestShow(t *testing.T) {
+	tk.MustExec("use test")
+	tk.MustExec("create table t(id int, abclmn int);")
+
+	defer tk.MustExec("drop table t;")
```
No need to remove tables, each MockStore will be created fresh
```suggestion 
```

mjonss reviewed (commented):
Looks good, with some comments.
There are also more tests that needs to update (since the SHOW INDEX output changed).
You can run all unit tests with `make dev` and you can run integration tests by first building the `tidb-server` binary with `make` and then `(cd tests/integrationtest ; ./run-tests.sh -s ../../bin/tidb-server)`

mjonss commented:
@samba-rgb I can fix the idc-jenkins-ci-tidb/mysql-test test failures. Can you fix the other failed tests?

samba-rgb commented:
@mjons Thank you for helping with my SQL test fix. I will now resume working on the fixes for integration and unit tests, along with addressing the comments mentioned above.

samba-rgb commented:
Hi @mjonss i added the  Global column in the integrationtest/r/*/*.result path
 example : tidb/tests/integrationtest/r/session/clustered_index.result
 but still in the check_dev its not reflecting 
 is their any place that i need to change/add
 
 BTW when i ran make dev no test case is running for me 
 find: md5sum: No such file or directory

mjonss commented:
> Hi @mjonss i added the Global column in the integrationtest/r/_/_.result path example : tidb/tests/integrationtest/r/session/clustered_index.result but still in the check_dev its not reflecting is their any place that i need to change/add
> 
> BTW when i ran` make dev` no test case is running for me find: md5sum: No such file or directory

@samba-rgb 
For the integration tests, you can simply re-record the result files, it seems like there is some formatting issues in your change, likely spaces instead of tabs. Try `./run-tests.sh -s ../../bin/tidb-server -r session/clustered_index` and it will overwrite the .result file, and you can try with `./run-tests.sh -s ../../bin/tidb-server -t session/clustered_index` to verify that it passes. 

I assume you are using a mac? md5sum is md5 in macOs..., you can also run only the unit tests with `make ut` IIRC.
Also if there are a few integration tests that fails on mac, it is due to floating point difference between arm and i386, so leave them as-is.

I will try to find which unit tests still needing to be updated.

mjonss commented:
pkg/ddl/db_change_test.go:857
        	Error:      	Received unexpected error:
        	            	need [[t 0 PRIMARY 1 c1 A 0 <nil> <nil>  BTREE   YES <nil> NO]], but got [[t 0 PRIMARY 1 c1 A 0 <nil> <nil>  BTREE   YES <nil> NO NO]]
        	Test:       	TestShowIndex

To run the test you must also enable failpoints, `make failpoint-enable`, which you then can disable before committing, `make failpoint-disable`
Maybe this patch file could help?
[56028.patch](https://github.com/user-attachments/files/17050432/56028.patch)
